src/main/java/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -480,25 +480,27 @@ private void parseDoc(Map<String, String> namespacePrefixesByUri, String lastSav
                 final String instanceId = instanceNodeIdStrs.get(instanceIndex);
                 final String instanceSrc = parseInstanceSrc(instance, lastSavedSrc);
 
-                // Only read in an external instance if its ID is used in the primary instance as an argument to an
+                // Only read in an secondary instance if its ID is used in the primary instance as an argument to an
                 // instance() call
-                if (instanceSrc != null && referencedInstanceIds.contains(instanceId)) {
-                    final ExternalDataInstance externalDataInstance;
-                    try {
-                        externalDataInstance = ExternalDataInstance.build(instanceSrc, instanceId);
-                    } catch (IOException | UnfullfilledRequirementsException | InvalidStructureException |
+                if (referencedInstanceIds.contains(instanceId)) {
```

Comment:
Moving this condition outwards somehow makes it easier to understand. I think that the `referencedInstanceIds` is a great addition that's giving us some unpredicted side-effect bonus :)

Reply:
I agree!